### PR TITLE
Prevent niche-optimization on futures to avoid triggering miri

### PR DESCRIPTION
**Pull request by Darksonn**
As suggested by @ralfjung at https://github.com/rust-lang/miri/issues/3780#issuecomment-2264999042

Initially pushing branch with just tests to ensure that they fail in CI without the fix.

Closes: #6729

### CODE REVIEW

**Comment by Darksonn** [comment]
Confirmed miri failure in CI:
```text
error: Undefined Behavior: trying to retag from <2420157> for SharedReadWrite permission at alloc824244[0x0], but that tag does not exist in the borrow stack for this location
   --> tokio/src/future/maybe_done.rs:95:17
    |
95  |                 *self.get_unchecked_mut().thing += ", world";
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                 |
    |                 trying to retag from <2420157> for SharedReadWrite permission at alloc824244[0x0], but that tag does not exist in the borrow stack for this location
    |                 this error occurs as part of two-phase retag at alloc824244[0x0..0x18]
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <2420157> was created by a Unique retag at offsets [0x0..0x18]
   --> tokio/src/future/maybe_done.rs:107:32
    |
107 |         let fut = async move { ThingAdder { thing: &mut thing }.await };
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: <2420157> was later invalidated at offsets [0x0..0x8] by a read access
   --> tokio/src/future/maybe_done.rs:8:1
    |
8   | / pin_project! {
9   | |     /// A future that may have completed.
10  | |     #[derive(Debug)]
11  | |     #[project = MaybeDoneProj]
...   |
22  | |     }
23  | | }
    | |_^
    = note: BACKTRACE (of the first span) on thread `future::maybe_d`:
    = note: inside `<future::maybe_done::miri_tests::ThingAdder<'_> as futures::Future>::poll` at tokio/src/future/maybe_done.rs:95:17: 95:48
note: inside closure
   --> tokio/src/future/maybe_done.rs:107:65
    |
107 |         let fut = async move { ThingAdder { thing: &mut thing }.await };
    |                                                                 ^^^^^
note: inside `<future::maybe_done::MaybeDone<{async block@tokio/src/future/maybe_done.rs:107:19: 107:72}> as futures::Future>::poll`
   --> tokio/src/future/maybe_done.rs:65:56
    |
65  |             MaybeDoneProj::Future { future } => ready!(future.poll(cx)),
    |                                                        ^^^^^^^^^^^^^^^
note: inside `future::maybe_done::miri_tests::maybe_done_miri`
   --> tokio/src/future/maybe_done.rs:115:20
    |
115 |         assert_eq!(fut.as_mut().poll(&mut ctx), Poll::Pending);
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside closure
   --> tokio/src/future/maybe_done.rs:102:25
    |
101 |     #[test]
    |     ------- in this procedural macro expansion
102 |     fn maybe_done_miri() {
    |                         ^
    = note: this error originates in the macro `$crate::__pin_project_enum_make_proj_method` which comes from the expansion of the attribute macro `test` (in Nightly builds, run with -Z macro-backtrace for more info)

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
```

File: tokio/src/future/maybe_done.rs [hunk 1]
```diff
@@ -69,3 +70,54 @@ impl<Fut: Future> Future for MaybeDone<Fut> {
         Poll::Ready(())
     }
 }
+
+// Test for https://github.com/tokio-rs/tokio/issues/6729
+#[cfg(test)]
```
Comment (hawkw) [inline tokio/src/future/maybe_done.rs:75]: should this also have `#[cfg(miri)]` on it? or do we also want these tests to run outside of miri?

Comment (Darksonn) [reply tokio/src/future/maybe_done.rs:75]: I mean, it doesn't really hurt to run it outside of miri too. :woman_shrugging:

Comment (hawkw) [reply tokio/src/future/maybe_done.rs:75]: fair enough!

**Review by mox692 — APPROVED** [review]
interesting, I didn't expect that the niche enum optimization was causing this miri error.👏

**Pull request merged**